src/mito2/src/worker/handle_flush.rs

```diff
@@ -14,81 +14,167 @@
+        if let Some(sender) = sender {
+            task.senders.push(sender);
+        }
+        if let Err(e) = self.flush_scheduler.schedule_flush(&region, task) {
+            error!(e; "Failed to schedule flush task for region {}", region.region_id);
+        }
     }
 
     /// On region flush job failed.
-    pub(crate) async fn handle_flush_failed(
-        &mut self,
-        _region_id: RegionId,
-        _request: FlushFailed,
-    ) {
-        // TODO(yingwen): fail all pending requests.
-        unimplemented!()
+    pub(crate) async fn handle_flush_failed(&mut self, region_id: RegionId, request: FlushFailed) {
+        self.flush_scheduler.on_flush_failed(region_id, request.err);
     }
 
     /// Checks whether the engine reaches flush threshold. If so, finds regions in this
     /// worker to flush.
-    pub(crate) fn maybe_flush_worker(&self) {
+    pub(crate) fn maybe_flush_worker(&mut self) {
         if !self.write_buffer_manager.should_flush_engine() {
             // No need to flush worker.
             return;
         }
+
         // If the engine needs flush, each worker will find some regions to flush. We might
         // flush more memory than expect but it should be acceptable.
-        self.find_regions_to_flush();
+        if let Err(e) = self.flush_regions_on_engine_full() {
+            error!(e; "Failed to flush worker");
+        }
     }
 
     /// Find some regions to flush to reduce write buffer usage.
-    pub(crate) fn find_regions_to_flush(&self) {
-        unimplemented!()
+    fn flush_regions_on_engine_full(&mut self) -> Result<()> {
+        let regions = self.regions.list_regions();
+        let now = current_time_millis();
+        let min_last_flush_time = now - self.config.auto_flush_interval.as_millis() as i64;
+        let mut max_mutable_size = 0;
+        // Region with max mutable memtable size.
+        let mut max_mem_region = None;
+
+        for region in &regions {
+            if self.flush_scheduler.is_flush_requested(region.region_id) {
+                // Already flushing.
+                continue;
+            }
+
+            let version = region.version();
+            let region_mutable_size = version.memtables.mutable_bytes_usage();
+            // Tracks region with max mutable memtable size.
+            if region_mutable_size > max_mutable_size {
+                max_mem_region = Some(region);
+                max_mutable_size = region_mutable_size;
+            }
+
+            if region.last_flush_millis() < min_last_flush_time {
+                // If flush time of this region is earlier than `min_last_flush_time`, we can flush this region.
+                let task = self.new_flush_task(region, FlushReason::EngineFull);
+                self.flush_scheduler.schedule_flush(region, task)?;
+            }
+        }
+
+        // Flush memtable with max mutable memtable.
+        if let Some(region) = max_mem_region {
```

Comment:
Looks like this method only flushes the region with max memtable? I wonder to what extent does this method help to reduce write buffer usage.

Reply:
We can optimize this later.